Tradução Webpacker > 3. Uso (#811)
campuscode/rails-guides-pt-BR: pt-BR/webpacker.md

The installation also calls the `yarn` package manager, creates a `package.json` file with a basic set of packages listed, and uses Yarn to install these dependencies.

Uso
-----

### Usando Webpacker para JavaScript

> Com Webpacker instalado, qualquer arquivo JavaScript no diretório `app/javascript/packs` será compilado para seu próprio arquivo de pacote padrão.

Reviewer: Humm não sei se deveriamos traduzir de fato **pack** para **pacote** aqui, tendo em vista que tudo ficaria no diretório `packs/`, pode ser que cause um pouco de confusão nos exemplos subsequentes, o que acham?
Acho que podemos deixar assim aqui e depois manter como **pack**
```suggestion
Com Webpacker instalado, qualquer arquivo JavaScript no diretório `app/javascript/packs` será compilado para seu próprio arquivo de _pack_ (pacote) padrão.
```

> Então, se você tem um arquivo chamado `app/javascript/packs/application.js`, Webpacker irá criar um pacote chamado `application`, e você pode adicionar isto para sua aplicação Rails com o código `<%= javascript_pack_tag "application" %>`. Com isso no lugar, em desenvolvimento, Rails irá recompilar o arquivo `application.js` a cada vez que é alterado, e você carrega a página que usa esse pacote. Normalmente, o arquivo no diretório `packs` real será um manifesto que carrega principalmente outros arquivos, mas também pode ter código JavaScript arbitrário.

Reviewer: ```suggestion
Então, se você tem um arquivo chamado `app/javascript/packs/application.js`, Webpacker irá criar um pacote chamado `application`, e você pode adicionar isto para sua aplicação Rails com o código `<%= javascript_pack_tag "application" %>`. Com isso no lugar, em desenvolvimento, Rails irá recompilar o arquivo `application.js` a cada vez que é alterado, e você carrega uma página que usa esse pacote. Normalmente, o arquivo no diretório `packs` real será um manifesto que carrega na sua maior parte outros arquivos, mas também pode ter código JavaScript arbitrário.
```

ActiveStorage.start()
```

Você precisará incluir um pacote que requeira esses pacotes para usá-los em seu aplicativo Rails.

> É importante notar que apenas os arquivos de entrada do Webpack devem ser colocados no diretório `app/javascript/packs`; O Webpack criará um gráfico de dependência separado para cada ponto de entrada, portanto, um grande número de pacotes aumentará a sobrecarga de compilação. O restante do código-fonte do seu ativo deve ficar fora desse diretório, embora o Webpacker não coloque nenhuma restrição ou faça sugestões sobre como estruturar seu código-fonte. Aqui está um exemplo:

Reviewer: ```suggestion
É importante notar que apenas os arquivos de entrada do Webpack devem ser colocados no diretório `app/javascript/packs`; O Webpack criará um gráfico de dependência separado para cada ponto de entrada, portanto, um grande número de pacotes aumentará a sobrecarga de compilação. O restante do seu código-fonte deve ficar fora desse diretório, embora o Webpacker não coloque nenhuma restrição ou faça sugestões sobre como estruturar seu código-fonte. Aqui está um exemplo:
```

      └── logo.svg
```

Normalmente, o próprio arquivo de pacote é em grande parte um manifesto que usa `import` ou `require` para carregar os arquivos necessários e também pode fazer alguma inicialização.

Se você quiser mudar esses diretórios, você pode ajustar o `source_path` (padrão `app/javascript`) e `source_entry_path` (padrão `packs`) no arquivo `config/webpacker.yml`.

> Dentro dos arquivos de origem, as instruções `import` são resolvidas em relação ao arquivo fazendo a importação, então `import Bar from "./foo"` encontra um arquivo `foo.js` no mesmo diretório que o arquivo atual, enquanto `import Bar from "../src/foo"` encontra um arquivo em um diretório irmão chamado `src`.

Reviewer: ```suggestion
Dentro dos arquivos de código fonte, as declarações `import` são resolvidas em relação ao arquivo fazendo a importação, então `import Bar from "./foo"` encontra um arquivo `foo.js` no mesmo diretório que o arquivo atual, enquanto `import Bar from "../src/foo"` encontra um arquivo em um diretório irmão chamado `src`.
```

Se você estiver usando um framework CSS, você pode adicioná-lo ao Webpacker seguindo as instruções para carregar o framework como um módulo NPM usando `yarn`, normalmente `yarn add <framework>`. A estrutura deve ter instruções sobre como importá-la para um arquivo CSS ou SCSS.

### Usando Webpacker para Assets Estáticos

A [configuração](https://github.com/rails/webpacker/blob/master/lib/install/config/webpacker.yml#L21) do Webpack padrão deve funcionar imediatamente para _assets_ estáticos.
A configuração inclui várias extensões de formato de arquivo de imagem e fonte, permitindo que o webpack as inclua no arquivo `manifest.json` gerado.

Com o webpack, os _assets_ estáticos podem ser importados diretamente em arquivos JavaScript. O valor importado representa a URL para o ativo. Por exemplo:

```javascript
import myImageUrl from '../images/my-image.jpg'

// ...
let myImage = new Image();
myImage.src = myImageUrl;
myImage.alt = "Eu sou uma imagem do pacote Webpacker";
document.body.appendChild(myImage);
```

> Se você precisar referenciar ativos estáticos do Webpacker a partir de uma visualização do Rails, os ativos precisam ser explicitamente exigidos dos arquivos JavaScript do pacote do Webpacker. Ao contrário do Sprockets, o Webpacker não importa seus ativos estáticos por padrão. O arquivo padrão `app/javascript/packs/application.js` tem um modelo para importar arquivos de um determinado diretório, que você pode descomentar para cada diretório em que deseja ter arquivos estáticos. Os diretórios são relativos a `app/javascript` . O template usa o diretório `images`, mas você pode usar qualquer coisa em `app/javascript`:

Reviewer: ```suggestion
Se você precisar referenciar _assets_ estáticos do Webpacker a partir de uma _view_ do Rails, os _assets_ precisam ser explicitamente requeridos nos arquivos JavaScript empacotados do Webpacker. Ao contrário do Sprockets, o Webpacker não importa seus _assets_ estáticos por padrão. O arquivo padrão `app/javascript/packs/application.js` tem um _template_ para importar arquivos de um determinado diretório, que você pode descomentar para cada diretório em que deseja ter arquivos estáticos. Os diretórios são relativos a `app/javascript` . O _template_ usa o diretório `images`, mas você pode usar qualquer coisa em `app/javascript`:
```

```
const images = require.context("../images", true)
const imagePath = name => images(name, true)
```

> _Assets_ estáticos serão enviados para um diretório em `public/packs/media`. Por exemplo, uma imagem localizada e importada em `app/javascript/images/my-image.jpg` será gerada em `public/packs/media/images/my-image-abcd1234.jpg`. Para renderizar uma tag de imagem para esta imagem em uma visualização do Rails, use `image_pack_tag 'media/images/my-image.jpg`.

Reviewer: ```suggestion
_Assets_ estáticos serão enviados para um diretório em `public/packs/media`. Por exemplo, uma imagem localizada e importada em `app/javascript/images/my-image.jpg` será gerada em `public/packs/media/images/my-image-abcd1234.jpg`. Para renderizar uma tag de imagem para esta imagem em uma _view_ do Rails, use `image_pack_tag 'media/images/my-image.jpg`.
```

Os _helpers_ do Webpacker ActionView para _assets_ estáticos correspondem aos _helpers_ de pipeline de _assets_ de acordo com a tabela a seguir:

|Helper ActionView | Helper Webpacker |
|------------------|------------------|
|favicon_link_tag  |favicon_pack_tag  |
|image_tag         |image_pack_tag    |

Além disso, o _helper_ genérico `asset_pack_path` seleciona a localização local de um arquivo e retorna sua localização do Webpacker para uso em _views_ do Rails.

Você também pode acessar a imagem referenciando diretamente o arquivo de um arquivo CSS em `app/javascript`.

### Webpacker em Engines Rails

> A partir da versão 6 do Webpacker, o Webpacker não é "consciente da _engine_", o que significa que o Webpacker não possui paridade de recursos com o Sprockets quando se trata de usar nas _engines_ do Rails.

Reviewer: ```suggestion
A partir da versão 6 do Webpacker, o Webpacker não é "consciente da _engine_", o que significa que o Webpacker não possui paridade de recursos com o Sprockets quando se trata de seu uso nas _engines_ do Rails.
```
